rfc: Define matching rules for form-urlencoded body (#99)
pact-foundation/roadmap: rfc/0000-define-matching-rules-for-form-urlencoded-body.md

---
name: define-matching-rules-for-form-urlencoded-body
started: 2024-09-06
pr: pact-foundation/roadmap#0000
tracking_issue: pact-foundation/roadmap#0000
---
## Summary

This RFC proposes a way to define matching rules for body of a `application/x-www-form-urlencoded` request.

## Motivation

Currently we can match the `application/x-www-form-urlencoded` request's body using matching rules. But there is no way to define
matching rules for the body to fully take advantage of this feature.

> The only a way to define body for a `application/x-www-form-urlencoded` request is to define it as-is.

Author: I is really confusing. It take a while for me to remember why I wrote this way.

> Currently we can match the `application/x-www-form-urlencoded` request's body using matching rules.

I mean pact-reference did have code to match form urlencoded's body located at https://github.com/pact-foundation/pact-reference/blob/master/rust/pact_matching/src/form_urlencoded.rs#L12. But currently there are no way to call this method `match_form_urlencoded`. So we can't use this feature (unless we implement this RFC).

Do you have suggestion on how can I rephrase this sentence?

That's because we can't define matching rules with this raw syntax. With the ability to define matching rules with the alternative syntax,
consumer will be able to send different values.

This example is focusing on request's body, but it also apply to response's body.

## Guide-level explanation

### With matching rules

When the RFC is implemented, the example usage would be:

```rust
let json = json!({
    "null": {
        "pact:matcher:type": "null"
    },
    "number": {
        "pact:matcher:type": "number",
        "value": 23.45
    },
    "string": {
        "pact:matcher:type": "type",
        "value": "example text"
    },
    "array": {
        "pact:matcher:type": "eachValue(matching(regex, 'value1|value2|value3|value4', 'value2'))",
        "value": ["value1", "value4"]
    }
});
pactffi_with_body(interaction, InteractionPart::Request, "application/x-www-form-urlencoded", json);
```

After the matching rules are extracted, the example body will be returned:

```query
number=23.45&string=example+text&array=value1&array=value4
```

This example body will be written into pact file, and will be used to send to provider to verify.

> > Note: Null will be removed

Author: `Null` is moved to unsupported list.

repeated keys and ordering are mentioned in the RFC.

        - Extract matching rules
        - Extract generators
        - Return example JSON body with example values e.g. `{ "key": "example value" }`
    - Example JSON body will be converted to example Form UrlEncoded body e.g. `key=example+value`
    - If JSON body has one of these cases, it will panic:
        - Boolean (true/false)
        - Object
        - Array of Arrays
        - Array of Objects
    - Register the interaction as normal with:
        - Extracted matching rules
        - Extracted generators
        - Example Form UrlEncoded body
        - Content type: `application/x-www-form-urlencoded`

### Unsupported syntax explaination

* Boolean: There is no standard way to represent boolean in query string. It can be 1/0, true/false or t/f. User need to define them explicitly.
* Object: There is no standard way to represent object in query string.
* Array of arrays: There is no standard way to represent array of arrays in query string.
* Array of objects: There is no standard way to represent array of objects in query string.

### Panic vs Warning

> To prevent matching rules and generators of unsupported syntax to be written into pact file, Rust core need to panic.

Author: panics are removed

> If we log a warning and ignore the unsupported syntax, matching rules and generators of ignored query params will be written to the pact file. This is undefined behavior.

Author: Error logs are raised. Generators are ignored. Matching rules of unsupported values are reverted.

### Generators

Generators will be written to pact file. Consumer test will failed with this warning log:

```
Generators only support JSON and XML
```

> User need to remove generators to make consumer test works.

Author: Yes, generators will be implemented later. Added a section for it.

---
name: define-matching-rules-for-form-urlencoded-body
started: 2024-09-06
pr: pact-foundation/roadmap#0000
tracking_issue: pact-foundation/roadmap#0000
---
## Summary

> This RFC proposes a way to define matching rules for body of a `application/x-www-form-urlencoded` request.

Reviewer: Minor suggestion:

```suggestion
This RFC proposes defining matching rules for bodies of type `application/x-www-form-urlencoded`.
```

Author: Updated

The values are ignored. Unsupported error messages are logged. The extracted example body will be empty.

### Special cases

- keys with no values (`a=&b=&c=`): can be achieved by define empty string value
- values with no keys (`=a&=b&=c`): can't be achieved because empty key is not supported and will be ignored
- repeated keys: (`a=1&a=2` or `a=2&a=1`): can be achieved by define array values
- special characters
    - `/`: will be encoded to `%2F`
    - `&`: will be encoded to `%26`
    - `?`: will be encoded to `%3F`
    - `=`: will be encoded to `%3D`
- only ampersands (`&&&`)
    - can't be defined by json
    - can be defined by raw syntax
    - these query strings can match it:
        - `&`
        - `&&`
        - `&&&`
        - ...
    - empty string can't match it
- only equals signs (`===`)
    - can't be defined by json
    - can be defined by raw syntax
    - these query strings can match it:
        - `===`
        - `=%3D%3D`
    - empty string can't match it
- no key and value (`=&=&=`): can't be achieved because empty key is not supported and will be ignored

> 

Reviewer 2: A couple of minor points here, and this is more in reference to the URL encoded RFC/specification I linked to above:

1. The `/` and `?` characters are typically _not_ special, even in a URL string. This means that an URL like `https://example.com/foo?a=?&?=/` has the query string `a=?&?=/`. This is an edge case which is often not properly handled, but it _is_ technically correct.
  - Having said that, to avoid confusion and possible implementation-specific variation, I would suggest that `/` and `%2F` both be allowed, and similarly with `?` and `%3F`. Effectively, we would just leave it up to the end-user to make the choice without Pact opining.
  - I agree that `&` and `=` still need to be escaped

For the `===` case, I would make it clear that its interpretation is ambiguous. I.e., there's no way to disambiguate parsing it as `{"=": "="}` or `{"": "=="}` or `{"==": ""}` (and even then, I'm not sure that is valid parsing... I should check the RFC again). I would specify that these should escape the `=` as above.

Lastly, you mention that empty string keys are not supported. Can you explain why? It would be ideal if we can support empty strings as we can then support `=a&=b&=c` as well as `=&=&=`.

Author: * `/` and `?` are currently encoded (both in key and value) in the current implementation of `process_json()` in pact-reference.
  * I think if we want to support these cases we need to change more code
  * I check with PHP's `http_build_query()` and it encoded them also i.e. `http_build_query(['/' => 'a', 'a' => '?', '?' => '/']) === '%2F=a&a=%3F&%3F=%2F'` 
* `===` can't be defined in json because theses cases are detected as plain text:
  * `{"=": "="}`
  * `{"": "=="}`
  * `{"==": ""}`
* Last time I tested empty key and it doesn't go well (matching rules fail always). Now I tested it again and it work fine, strange. So I support empty key again.
